osemgrep support for --scan-unknown-extensions

mjambon opened the pull request:
So far, the `Find_targets.conf` config was used only for generic target filtering that was independent of language detection. It is now also used for language detection. The previously unused option `scan_unknown_extensions` was renamed `always_select_explicit_targets` for clarity and now causes language detection to be bypassed for files that are explicit targets.

The set of explicit targets needed somehow to be passed around. I put it as a new field in `Find_targets.conf`. Alternatively, I could have represented each target as a pair (path, is_explicit) but it seemed heavier.

Other: I added a pytest test that exercises the combination of `-e` with `--lang` and `--scan-unknown-extensions`. For the test to pass with osemgrep, I had to populate the `"message"` field of JSON results with the value of the `-e` pattern as it is done by pysemgrep.

Commit:
Convert Core_scan.ml to use the Logs library

Commit:
Simplify

Commit:
Add osemgrep support for --scan-unknown-extensions
This requires tracking which target file is an explicit target and
passing the relevant option around to the point where language
detection occurs so it can be bypassed.

aryx commented:
lots of CI failures :(
otherwise looks good

Commit:
Update test snapshots after one target file was added

Commit:
Add missing snapshot

Commit:
Fix more snapshots

Commit:
Add OCaml tests for file targeting. They need fixing.

aryx commented on cli/tests/e2e/snapshots/test_rule_parser/test_rule_parser_cli_pattern/error.txt:
```diff
@@ -3,7 +3,7 @@
 │ Scan Status │
 └─────────────┘
-  Scanning 23 files tracked by git with 1 Code rule:
+  Scanning 24 files tracked by git with 1 Code rule:
```
why we suddently have more files scanned in the snapshot? Did you modify pysemgrep file targeting?

mjambon replied:
I added a file to `cli/tests/e2e/targets/basic/` which is used as the default scanning root in e2e tests.

aryx commented on src/osemgrep/cli_scan/Scan_CLI.ml:
```diff
@@ -245,15 +230,18 @@ in a git repository.
       (spf
-         {|If true, explicit files will be scanned using the language specified
-in --lang. If --skip-unknown-extensions, these files will not be scanned.
+         {|If true, target files specified directly on the command line
```
Far clearer.

aryx replied:
cc @armchairlinguist who I think had some beef with this flag in the past.

armchairlinguist replied:
If I did, I don't recall it! Sorry 😕

aryx commented on src/osemgrep/cli_scan/Scan_subcommand.ml:
```diff
@@ -535,9 +534,14 @@ let run_scan_files (_caps : < Cap.stdout >) (conf : Scan_CLI.conf)
             Some (file_match_results_hook conf filtered_rules) )
       | { output_conf; _ } -> (output_conf.output_format, None)
     in
+    (* TODO: What is this wrapping for? It makes things really hard to
```
true

aryx commented on src/osemgrep/cli_scan/Scan_subcommand.ml:
```diff
@@ -535,9 +534,14 @@ let run_scan_files (_caps : < Cap.stdout >) (conf : Scan_CLI.conf)
     let scan_func = mk_scan_func conf file_match_results_hook errors in
     (* step 3': call the engine! *)
     let exn_and_matches =
+      (* TODO: this long code block should not be here!
```
true. Maybe @ihji can cleanup a bit this code.

aryx commented on src/targeting/Find_targets.ml:
```diff
@@ -36,6 +36,28 @@ module Out = Semgrep_output_v1_t
 
 module Fppath_set = Set.Make (Fppath)
 
+module Explicit_targets = struct
```
I'm not a huge fan of nested modules. I'd rather move this code in a separate Explicit_targets.ml file.

mjambon replied:
noted.

aryx reviewed (approved):
Very nice! Just need to fix the valid CI regressions (I think that semgrep-pro is failing
is normal since you change the scan_func_for_osemgrep thing.

Commit:
Add a module Rfpath to hold both realpath and original path + use it to
represent scanning roots. This revealed lots of confusing code where
discovered targets may be used as scanning roots again.

Commit:
Add machinery to allow running git from outside of git or some folder
other than the repo root.

Commit:
Test and fix the new Git_wrapper.ls_files_relative that allows listing
files from outside the git repo.

Commit:
Add test snapshots

Commit:
Remove the extra tests for Ppath that have not been exercised in months,
depend on git_wrapper, fail, and are complicated.

Commit:
Add tests for git project root finding (WIP)

Commit:
Force some logging to be able to locate the source of mysterious exit statuses

Commit:
Fix test that was failing to a chdir occurring after calling getcwd.

Commit:
Update test snapshots

Commit:
Add Rfpath.is_valid

Commit:
Mark 5 tests as passing with osemgrep

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Fix post-merge issues

Commit:
Fix bad masking regexp

Commit:
Provide equality function for Exit_code

Commit:
Add some support for stdin as '-' and named pipes on the command line.
(untested)

Commit:
Assume --scan-unknown-extensions in osemgrep when one of the explicit
targets is stdin ('-') or a named pipe like pysemgrep does.
This doesn't solve all the problems, but appears to work partially. WIP

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Fix post-merge type errors

Commit:
Clean-up

Commit:
Rename hook to be consistent with the type name

Commit:
Change the type of the scanning roots on the semgrep-core command line
to clarify things in semgrep-pro

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Avoid consuming stdin ('-') or named pipes if the command is going to be
forwarded as-is to pysemgrep.

Commit:
Produce an error message instead of 'assert false'

aryx commented:
conflicts

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Provide functions to log paths that can't be converted to physical paths

Commit:
Merge commit '63d214885' into martin/scan-unknown

Commit:
Update testo

Commit:
Fix package name

Commit:
Merge commit 'cdafbb1bc5fb6cb92038acc93e20d13575327f53' into martin/scan-unknown

Commit:
Change the type of scanning roots back to an alias for Fpath.t after
realizing that they may contain symlinks that should not be resolved
before determining the project root.

Commit:
Merge commit '0fc51a30ae0b4ea870ed544822b5c2e4d65a7d23' into martin/scan-unknown

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Fix type error in test code

Commit:
Change the type of scanning roots to Scanning_root.t for semgrep-core

Commit:
Clean up

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Fix type errors

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

Commit:
Merge remote-tracking branch 'origin/develop' into martin/scan-unknown

mjambon commented:
https://github.com/semgrep/semgrep-proprietary/pull/1317 is looking good. Merging!

aryx commented on TCB/TCB.ml:
```diff
@@ -334,9 +334,13 @@ let ( @ ) = ( @ )
  *)
 
+(*
+I don't know how to use stdin if it's shadowed -- Martin
```
That's the point. You have to qualify with UStdlib.stdin so it's explicit you're using
the unsafe part of the stdlib (U means unsafe). I'll add a comment about it in TCB.ml

aryx commented on libs/commons/UCmd.ml:
```diff
@@ -9,6 +9,21 @@ open Common
+   Let's not log environment variables because they may contain sensitive
+   secrets.
+*)
+let log_command cmd =
```
nice

aryx commented on src/core/Scanning_root.mli:
```diff
@@ -0,0 +1,22 @@
+(*
```
Can this file be moved in src/targeting/ instead?

aryx commented on src/core_cli/Core_CLI.ml:
```diff
@@ -425,7 +425,13 @@ let options caps actions =
+        (fun s ->
+          let path = Fpath.v s in
+          (*
+        Printf.eprintf "-rules:\n%s\n%!" (UFile.read_file path);
```
remove comment or use Logs

aryx commented on src/core_cli/Core_CLI.ml:
```diff
@@ -435,7 +441,13 @@ let options caps actions =
+      Arg.String
+        (fun s ->
+          let path = Fpath.v s in
+          (*
```
same

aryx commented on src/core_scan/Core_scan_config.ml:
```diff
@@ -44,7 +44,8 @@ type t = {
   equivalences_file : Fpath.t option;
   lang : Xlang.t option;
-  roots : Fpath.t list;
+  (* Scanning roots. They are mutually exclusive with target_source! *)
```
I don't understand this comment.

aryx commented on src/osemgrep/cli_publish/Test_publish_subcommand.ml:
```diff
@@ -150,28 +140,14 @@ let test_publish (caps : < Cap.network ; Cap.stdout >) () =
               [| "semgrep-publish"; "--visibility=public"; !!valid_target |]
           in
-          assert (exit_code =*= Exit_code.fatal);
-
```
why remove those comments? They are basically TODOs I think.

aryx commented on src/osemgrep/core/Exit_code.ml:
```diff
@@ -4,57 +4,109 @@
+   Otherwise, it's like looking for a needle in haystack.
+*)
+let with_log ~__LOC__:loc x =
+  Logs.info (fun m ->
```
probably better as a Logs.debug

aryx reviewed (commented):
Nice. Lots of good stuff. Would be better though in the futur to split in different PRs.
one for Exit_code and the ~__LOC, one for the Base comment, one for rrpath types, etc.

aryx commented:
I'm reverting this. We have 9 error regressions:
```

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
ERROR tests/e2e/test_exclude_include.py::test_exclude_include_verbose_sorted_1
ERROR tests/e2e/test_exclude_include.py::test_exclude_include_verbose_sorted_2
ERROR tests/e2e/test_output.py::test_debug_experimental_rule - Failed: Snapsh...
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_file_parser__failure__error_messages[settings0]
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_file_parser__failure__error_messages[settings1]
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_rule_parser__failure__error_messages[settings0]
ERROR tests/e2e/test_misc.py::test_max_memory - Failed: Snapshot directory wa...
ERROR tests/e2e/test_test.py::test_parse_errors - Failed: Snapshot directory ...
ERROR tests/e2e/test_misc.py::test_timeout_threshold - Failed: Snapshot direc...
= 1565 passed, 1 skipped, 2 xfailed, 12569 warnings, 9 errors in 226.87s (0:03:46) =
make: *** [Makefile:52: test-for-ci] Error 1
Error: Process completed with exit code 2.`
```

aryx commented:
Maybe the opportunity as you restore this PR to add it piece by piece, like the core_run vs core_scan rename, 
the Rfpath introduction, etc., so it's more incremental.

mjambon commented:
The run https://github.com/semgrep/semgrep/actions/runs/8120059773/job/22196856080 failed but I didn't see it on Friday. Also, I haven't been able to reproduce these failures today.
